### PR TITLE
getDirectServiceProxyFromUrlString added and tested

### PR DESCRIPTION
**Pull request by silversoul93**
#### Checklist

- [x] your branch will not cause merge conflict with `master`
- [x] run `npm test`
- [x] tests are included
- [x] the documentation is updated or intregrated accordingly with your changes
- [x] the code will follow the lint rules and style of the project
- [x] you are not committing extraneous files or sensible data

### CODE REVIEW

File: tests/getServiceProxy.test.js [hunk 1]
```diff
@@ -73,3 +82,103 @@ tap.test('getServiceProxy available for testing', async t => {
...
+
+tap.test('getDirectServiceProxyFromUrlString available for testing - https url', async t => {
+  nock.disableNetConnect()
+  t.tearDown(() => {
+    nock.enableNetConnect()
+  })
+
+  const RETURN_MESSAGE = 'OK'
+  const customProxy = getDirectServiceProxyFromUrlString(MY_AWESOME_SERVICE_PROXY_HTTPS_URL)
+  const awesomeHttpsServiceScope = nock(`${MY_AWESOME_SERVICE_PROXY_HTTPS_URL}:443`)
```
Comment (theliuk) [inline tests/getServiceProxy.test.js:115]: I don't get why you want to add the port by means of interpolation; couldn't it be part of the constant `MY_AWESOME_SERVICE_PROXY_HTTPS_URL` in the first place? after all, the port is part of the URL.

**Comment by theliuk** [comment]
LGTM

File: tests/getServiceProxy.test.js (hunk 1 above)
Comment (silversoul93) [reply tests/getServiceProxy.test.js:115]: I did it just to be sure that the default ports are still working without specifying them.
I append the port into the nock only, but not into the `getDirectServiceProxyFromUrlString` function parameter.

**Comment by davidebianchi** [comment]
Thanks for the PR!

Can you explain me the use case you imagine for this method and why this should be used by a user?

**Comment by silversoul93** [comment]
> Thanks for the PR!
> 
> Can you explain me the use case you imagine for this method and why this should be used by a user?

@davidebianchi this method is useful when a user needs to get a direct service proxy by using a complete url, that includes protocol and port.
Using the `getDirectServiceProxy` function with an url like `http://my-custom-url:3000` will cause an error using whatever function provided by the proxy.

For example:
```javascript
const urlString = 'http://my-custom-url:3000'
const myProxy = service.getDirectServiceProxy(urlString)
const result = await myProxy.get('/searchUser') // --> error "getaddrinfo ENOTFOUND http"
```

**Comment by theliuk** [comment]
I devised this simple method since, most of the times, I want to specify a URL rather than a simple hostname. getDirectServiceProxy was initially introduced to call internal services by name, and that’s fine, but if you want to contact external service, a URL is much better.

Frengo took this simple function which was spreading across many projects and wanted to make it available out of the box.

Sent from my iPhone

On 6 Feb 2021, at 17:57, Davide Bianchi <notifications@github.com> wrote:

﻿

Thanks for the PR!

Can you explain me the use case you imagine for this method and why this should be used by a user?

**Comment by davidebianchi** [comment]
> > Thanks for the PR!
> > 
> > Can you explain me the use case you imagine for this method and why this should be used by a user?
> 
> @davidebianchi this method is useful when a user needs to get a direct service proxy by using a complete url, that includes protocol and port.
> Using the `getDirectServiceProxy` function with an url like `http://my-custom-url:3000` will cause an error using whatever function provided by the proxy.
> 
> For example:
> ```javascript
> const urlString = 'http://my-custom-url:3000'
> const myProxy = service.getDirectServiceProxy(urlString)
> const result = await myProxy.get('/searchUser') // --> error "getaddrinfo ENOTFOUND http"
> ```

Ok, clear. Thank you! What do you think about support this functionality directly in getDirectServiceProxy function? 
It could be easily implemented checking if the URL starts with http/https protocol.

Do you see any downside?

**Comment by silversoul93** [comment]
> > > Thanks for the PR!
> > > Can you explain me the use case you imagine for this method and why this should be used by a user?
> > 
> > 
> > @davidebianchi this method is useful when a user needs to get a direct service proxy by using a complete url, that includes protocol and port.
> > Using the `getDirectServiceProxy` function with an url like `http://my-custom-url:3000` will cause an error using whatever function provided by the proxy.
> > For example:
> > ```js
> > const urlString = 'http://my-custom-url:3000'
> > const myProxy = service.getDirectServiceProxy(urlString)
> > const result = await myProxy.get('/searchUser') // --> error "getaddrinfo ENOTFOUND http"
> > ```
> 
> Ok, clear. Thank you! What do you think about support this functionality directly in getDirectServiceProxy function?
> It could be easily implemented checking if the URL starts with http/https protocol (or even better use the URL for every entry should also work).
> 
> Do you see any downside?

I see just one potential downside. The `getDirectServiceProxy` function never throws, but the `getDirectServiceProxyFromUrlString` function does.
It's a problem if the `getDirectServiceProxy` function start

**Comment by silversoul93** [comment]
> > > Thanks for the PR!
> > > Can you explain me the use case you imagine for this method and why this should be used by a user?
> > 
> > 
> > @davidebianchi this method is useful when a user needs to get a direct service proxy by using a complete url, that includes protocol and port.
> > Using the `getDirectServiceProxy` function with an url like `http://my-custom-url:3000` will cause an error using whatever function provided by the proxy.
> > For example:
> > ```js
> > const urlString = 'http://my-custom-url:3000'
> > const myProxy = service.getDirectServiceProxy(urlString)
> > const result = await myProxy.get('/searchUser') // --> error "getaddrinfo ENOTFOUND http"
> > ```
> 
> Ok, clear. Thank you! What do you think about support this functionality directly in getDirectServiceProxy function?
> It could be easily implemented checking if the URL starts with http/https protocol (or even better use the URL for every entry should also work).
> 
> Do you see any downside?

I just see one possible downside. The `getDirectServiceProxy` function actually never throws, but the `getDirectServiceProxyFromUrlString` does.
Do you see any problem if the `getDirectServiceProxy` function starts throw with this modification?
I think that the function should throw if the URL object creation fails, for example:
```javascript
function getDirectServiceProxy(serviceName = '', options = {}) {
  if (serviceName.startsWith('http://') || serviceName.startsWith('https://')) {
    try {
      const { hostname, protocol, port } = new URL(serviceName)
      // ....... other instructions
    } catch(urlCreationError) {
      throw new Error(`Error during the creation of the URL: ${urlCreationError.message}`)
    }
  }
}
```

**Comment by davidebianchi** [comment]
Now the `getDirectServiceProxy` does not throw because the `new URL` [is created based on the fake URL `http://example.com`](https://github.com/mia-platform/custom-plugin-lib/blob/eaa569ecc7cb938b2f0c485a17526d86fbc54f12/lib/serviceBuilder.js#L141) and overwrite it later adding the correct hostname passed by the user.
The resulting URL currently passing an absolute url starting with `http:` is `http://http/` (from here the error you have as result of the request).

Checking that the URL is absolute, you should never obtain a TypeError unless special characters are present, for example, in the hostname.

To avoid to throw, the implementation could be similar to the existent, and if the URL is not valid could be set to a default hostname without throw and throw once the request is executed.  
It could be ok to throw if service is created at deploy time (so the url is checked before service starts), instead if it is created from request I suppose it is not good to throw (but maybe better than call something potentially wrong)

**Comment by silversoul93** [comment]
> Now the `getDirectServiceProxy` does not throw because the `new URL` [is created based on the fake URL `http://example.com`](https://github.com/mia-platform/custom-plugin-lib/blob/eaa569ecc7cb938b2f0c485a17526d86fbc54f12/lib/serviceBuilder.js#L141) and overwrite it later adding the correct hostname passed by the user.
> The resulting URL passing an absolute url which start with `http:` is `http://http/` (from here the error you have as result of the request).
> 
> Checking that the URL is absolute, you should never obtain a TypeError unless special characters are present, for example, in the hostname.
> 
> To avoid to throw, the implementation could be similar to the existent, and if the URL is not valid could be set to a default hostname without throw and throw once the request is executed.
> It could be ok to throw if service is created at deploy time (so the url is checked before service starts), instead if it is created from request I suppose it is not good to throw (but maybe better than call something potentially wrong)

@davidebianchi I applied all your suggestions. Now the `getDirectServiceProxy` function:
 - can receive a complete url as `serviceName` parameter, like `http://my-awesome-service:3000`
 - does not throw if `new URL(serviceName)` fails, but returns the proxy built with `serviceBuilder(serviceName, {}, options)`

File: docs/HTTPClient.md [hunk 2]
```diff
@@ -8,15 +8,15 @@ You can get a proxy calling these methods both on `Request`(the first argument o
...
+    * `protocol` - a string that identifies the protocol to use (only `http` and `https` are supported, the default value is `http`)
+    * `headers` - an object that represents the set of headers to send to the service
+    * `prefix` - a string representing the prefix of the service call path
     * `timeout` - set a request timeout
     * `agent` - set a custom node agent
-* `getDirectServiceProxy(serviceName, options)` - returns a direct proxy to the service.
-  * `serviceName` - The name of the service to call. You can't specify the port here, you have to do in `options.port`.
-  * `options` - The same options described above.
+* `getDirectServiceProxy(serviceName, options)` - returns a direct proxy to the service
+  * `serviceName` - The name of the service to call. You can pass just the hostname, without protocol and port (that you can specify into the `options` field), or a complete url string
```
[the remark this answers is not in the document]
Comment (silversoul93) [reply docs/HTTPClient.md:18]: Changed, I'll change it on other files too

**Comment by silversoul93** [comment]
@davidebianchi I updated the `getDirectServiceProxy` added to the request to use the new logic.
I rebased my branch too, following to your update on master.

**Comment by silversoul93** [comment]
@davidebianchi @fredmaggiowski up. I just rebased this branch following to the @ilteoood pull request

File: docs/HTTPClient.md [hunk 3]
```diff
@@ -8,15 +8,15 @@ You can get a proxy calling these methods both on `Request`(the first argument o
...
+    * `protocol` - a string that identifies the protocol to use (only `http` and `https` are supported, the default value is `http`)
+    * `headers` - an object that represents the set of headers to send to the service
+    * `prefix` - a string representing the prefix of the service call path
     * `timeout` - set a request timeout
     * `agent` - set a custom node agent
-* `getDirectServiceProxy(serviceName, options)` - returns a direct proxy to the service.
-  * `serviceName` - The name of the service to call. You can't specify the port here, you have to do in `options.port`.
-  * `options` - The same options described above.
+* `getDirectServiceProxy(serviceNameOrURL, options)` - returns a direct proxy to the service
+  * `serviceNameOrURL` - The name of the service to call. You can pass just the hostname, without protocol and port (that you can specify into the `options` field), or a complete url string
```
Comment (fredmaggiowski) [inline docs/HTTPClient.md:18]: When providing a fully qualified URL how are treated port and protocol options? Are they ignored? Can we express this in the doc?

File: docs/HTTPClient.md [hunk 4]
```diff
@@ -44,7 +44,7 @@ The params to be passed to these functions are:
   * a [Buffer](https://nodejs.org/api/buffer.html#)
   * one [Stream](https://nodejs.org/api/stream.html)
 * `querystring` - optional, an object that represents the querystring.
-* `options` - optional, an object that admits all the`options` listed above for the `getServiceProxy` and`getDirectServiceProxy` methods (which will eventually be overwritten), plus the following fields:
+* `options` - optional, an object that admits all the`options` listed above for the `getServiceProxy` and `getDirectServiceProxy` methods (which will eventually be overwritten), plus the following fields:
```
Comment (fredmaggiowski) [inline docs/HTTPClient.md:47]:
```suggestion
* `options` - optional, an object that admits all the `options` listed above for the `getServiceProxy` and `getDirectServiceProxy` methods (which will eventually be overwritten), plus the following fields:
```

File: index.js [hunk 5]
```diff
@@ -122,15 +122,30 @@ function getBaseOptionsDecorated(headersKeyToProxy, baseOptions, headers) {
...
+  } catch (error) {
+    return serviceBuilder(serviceNameOrURL, serviceHeaders, options)
+  }
 }
 
-function getDirectlyServiceBuilderFromService(serviceName, baseOptions = {}) {
-  return serviceBuilder(serviceName, {}, baseOptions)
+function getDirectlyServiceBuilderFromService(serviceNameOrURL, baseOptions = {}) {
+  try {
+    const proxy = getDirectServiceProxyFromUrlString(serviceNameOrURL, {}, baseOptions)
```
Comment (fredmaggiowski) [inline index.js:144]: why not directly return it?

**Review by fredmaggiowski — CHANGES_REQUESTED** [review]
some minor changes

File: index.js (hunk 5 above)
Comment (silversoul93) [reply index.js:144]: Nice question, I don't know 😂 fixed

File: docs/HTTPClient.md (hunk 3 above)
Comment (silversoul93) [reply docs/HTTPClient.md:18]: Port and protocol are correctly used. I added this detail to the documentation

**Review by davidebianchi — APPROVED** [review]

**Review by fredmaggiowski — APPROVED** [review]

**Pull request merged**